File: prob1.java
// Time Complexity : O(log(m*n))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

public class prob1 {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix==null||matrix.length==0)
            return false;

        int low = 0;
        int high = (matrix.length*matrix[0].length)-1;
        while(low<=high){
            int mid = low - (low-high)/2;
            int rowNumber = mid/matrix[0].length;
            int columnNumber = mid%matrix[0].length;
            if(matrix[rowNumber][columnNumber]==target)
                return true;
            else if(matrix[rowNumber][columnNumber]<target)
                low = mid+1;
            else
                high = mid-1;
        }
        return false;
    }
}

File: Prob1.py
# // Time Complexity : O(log n)
# // Space Complexity : O(1)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
# // Your code here along with comments explaining your approach in three sentences only
# Main logic is that at least 1 half of the array (while inclduing mid) has to sorted, so find which half is sorted and then do binary search to find the  target element.
# If not do binary search in the other half (without mid) and find the target element.
# Else return -1


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums)==0: 
            return -1
        low=0
        high=len(nums)-1

        while low<=high:
            mid=low+(high-low)//2
            if nums[mid]==target:
                return mid
            if nums[low]<=nums[mid]: #if left sorted
                if nums[low]<=target and nums[mid]>target:
                    high=mid-1
                else:
                    low=mid+1
            else: #if right sorted
                if nums[mid]<target and nums[high]>=target:
                    low=mid+1
                else:
                    high=mid-1
        return -1


